src/reporting/clients/ODataClient.ts

```diff
@@ -0,0 +1,88 @@
+import { Collection, getEntityCollectionPage } from "../iterators/IteratorUtil";
+import { OperationsBase } from "../OperationsBase";
+import type { IOdataClient } from "./IODataClient";
+
+export class ODataClient extends OperationsBase implements IOdataClient{
+  public async getODataReport(accessToken: AccessToken, reportId: string): Promise<ODataResponse> {
+    const url = `${this.basePath}/odata/${encodeURIComponent(reportId)}`;
+    const requestOptions: RequestInit = this.createRequest("GET", accessToken);
+    return this.fetchJSON<ODataResponse>(url, requestOptions);
+  }
+
+  public async getODataReportEntityPage(accessToken: AccessToken, reportId: string, odataItem: ODataItem, sequence: number): Promise<ODataEntityResponse> {
+    const segments = odataItem.url.split("/");  // region, manifestId, entityType
+    if (segments.length !== 3) {
+      throw new RequiredError(
+        "odataItem",
+        "odata item was invalid when calling getODataReportEntityPage.",
+      );
+    }
+    const url = `${this.basePath}/odata/${encodeURIComponent(reportId)}/${odataItem.url}?sequence=${encodeURIComponent(sequence)}`;
+    const requestOptions: RequestInit = this.createRequest("GET", accessToken);
+    return this.fetchJSON(url, requestOptions);
+  }
+
+  public async getODataReportEntities(accessToken: AccessToken, reportId: string, odataItem: ODataItem): Promise<Array<ODataEntityValue>> {
+    const segments = odataItem.url.split("/");  // region, manifestId, entityType
+    if (segments.length !== 3) {
+      throw new RequiredError(
+        "odataItem",
+        "odata item was invalid when calling getODataReportEntities.",
+      );
+    }
+    const reportData: Array<ODataEntityValue> = [];
+    const oDataReportEntitiesIt = this.getODataReportEntitiesIterator(accessToken, reportId, odataItem);
+    for await(const oDataReportEntity of oDataReportEntitiesIt) {
+      reportData.push(oDataReportEntity);
+    }
+    return reportData;
+  }
+
+  public getODataReportEntitiesIterator(accessToken: AccessToken, reportId: string, odataItem: ODataItem): EntityListIterator<ODataEntityValue> {
+    const segments = odataItem.url.split("/");  // region, manifestId, entityType
+    if (segments.length !== 3) {
+      throw new RequiredError(
+        "odataItem",
+        "odata item was invalid when calling getODataReportEntitiesIterator.",
+      );
+    }
+    const url = `${this.basePath}/odata/${encodeURIComponent(reportId)}/${odataItem.url}`;
+    const request = this.createRequest("GET", accessToken);
+    return new EntityListIteratorImpl(async () => getEntityCollectionPage<ODataEntityValue>(
+      url,
+      async (nextUrl: string): Promise<Collection<ODataEntityValue>> => {
+        const response: ODataEntityResponse = await this.fetchJSON<ODataEntityResponse>(nextUrl, request);
+        const link: PagedResponseLinks = {
+          self: {
+            href: url,
```

Comment:
I think the `self` link might be incorrect here. Self link is the url of the request that you get the response for. Let's say you make a request to `https://abc.com/something`, then the `self` url in the response should be `https://abc.com/something`.

In case of this method, the self link should be set to `nextUrl` because a couple lines above you have a call to `fetchJSON` where you pass in the `nextUrl` and this method handles the response of that request.

Reply:
Changed to nextUrl